doc/source/scripts/base.py

```diff
@@ -18,87 +70,73 @@ class TableWriter(object):
-    preprocess = []
-    postprocess = []
-    sort = True
+    def __init__(
+        self,
+        filename: str = "",
+        include_table: Optional[str] = None,
+        sort: bool = True,
+        input_items: Iterable[Any] = set(),
+        column_spec: list[tuple[str, Callable[..., str]]] = [],
+        lines: list[list[str]] = [],
+    ):
+        if column_spec:
+            assert input_items
+        assert (column_spec and not lines) or (lines and not column_spec)
+        stem = os.getcwd().split("source")[0]
+        self.path = os.path.join(stem, "source", filename)
+        self.filename = filename
+        self.include_table = include_table
+        self.sort = sort
+        self.input_items = input_items
+        self.column_spec = column_spec
+        self.lines = lines
+        self._df = pd.DataFrame()
 
-    def __getattr__(self, key: str) -> list:
-        return self.fields[key]
+    @property
+    def headers(self) -> list[str]:
+        return [column_name for column_name, _ in self.column_spec]
 
-    def __init__(self, *args, **kwargs):
-        stem = os.getcwd().split("source")[0]
-        self.path = os.path.join(stem, "source", self.filename)
-        self.fields = defaultdict(list)
+    @property
+    def fields(self) -> pd.DataFrame:
+        return self._df
 
-        parent_directory = pathlib.Path(self.path).parent
-        parent_directory.mkdir(exist_ok=True, parents=True)
-        self.get_lines(*args, **kwargs)
-        self.write_table()
+    def generate_lines_and_write_table(self) -> None:
+        df = _generate_table(
+            input_items=self.input_items,
+            column_spec=self.column_spec,
+        )
 
-    def _run_method(self, method, *args, **kwargs):
-        sanitized = self.sanitize_name(method)
-        meth = getattr(self, sanitized)
-        val = meth(*args, **kwargs)
-        self.fields[method].append(val)
-        return val
-
-    @staticmethod
-    def sanitize_name(name):
-        return "_" + name.replace(" ", "_").replace("/", "_").lower()
-
-    def get_lines(self, *args, **kwargs):
-        lines = []
-        for items in self._set_up_input():
-            try:
-                lines.append(self.get_line(*items))
-            except TypeError:  # one argument
-                lines.append(self.get_line(items))
-        if self.sort:
-            lines = sorted(lines)
+        lines = df.values.tolist()
+        lines = sorted(lines) if self.sort else lines
         self.lines = lines
+        self._df = df
+        self.write_table()
+
+    def write_table(self) -> None:
+        write_table(
+            path=self.path,
+            headers=self.headers,
+            lines=self.lines,
+            include_table=self.include_table,
+        )
+
+
+# ==== HELPER FUNCTIONS ==== #
+
+
+def sphinx_class(*, klass: Type[Any], tilde: bool = True) -> str:
+    prefix = "~" if tilde else ""
+    return ":class:`{}{}.{}`".format(prefix, klass.__module__, klass.__name__)
```

Comment:
```suggestion
    return f":class:`{prefix}{klass.__module__}.{klass.__name__}`"
```

Would f-strings be easier here?